### PR TITLE
More specific exception classes

### PR DESCRIPTION
**Pull request by casperisfine**
Take over: https://github.com/brianmario/mysql2/pull/870
Partially fix: https://github.com/brianmario/mysql2/issues/404

I've just implemented 2 of them to showcase what it could look like. I can go over all error codes if dimmed necessary.

(Sorry for opening a second PR, but long story short I lost access to the first fork I did).

@sodabrew does this look good to you?

cc @rafaelfranca & @sgrif

### CODE REVIEW

**Comment by sodabrew** [comment]
This looks good to me. Will need a rebase since the Ruby 1.9-syntax changes landed in another PR. Could you run down the unit test failures?

**Comment by casperisfine** [comment]
@sodabrew sorry I should have checked CI after my rebase.

I fixed the test suite. Should be good to go now.

File: lib/mysql2/error.rb [hunk 1]
```diff
@@ -8,25 +8,60 @@ class Error < StandardError
...
-      @server_version ||= nil
+    def initialize(msg, server_version = nil, error_number = nil, sql_state = nil)
+      @server_version = server_version
+      @error_number = error_number
+      @sql_state = sql_state ? sql_state.encode(ENCODE_OPTS) : nil
 
       super(clean_message(msg))
     end
 
     def self.new_with_args(msg, server_version, error_number, sql_state)
```
Comment (sodabrew) [inline lib/mysql2/error.rb]: I remember there was a reason for all of this, but I don't remember what that reason was 🤕

**Comment by sodabrew** [comment]
Thanks very much for the rebase and updating the unit tests!

File: lib/mysql2/error.rb (hunk 1 above)
Comment (casperisfine) [reply lib/mysql2/error.rb]: I remember git blaming it and figuring out that it wasn't required anymore. Let me blame again.

Comment (casperisfine) [reply lib/mysql2/error.rb]: Hum, don't remember why I dimmed it unnecessary https://github.com/brianmario/mysql2/pull/545

I still think it's better this way, but I can remove that change if you prefer.

Comment (sodabrew) [reply lib/mysql2/error.rb]: I moved the arguments into the constructor in f9eba44b0519580d2cd52b51d2b902f45186f427 but moved them out to a separate method in 1d4de8963c71ce88772a0e27883066160c3dbbd4

**Comment by sodabrew** [comment]
LGTM. Do you want to wait for feedback from the other folks you tagged in the OP?

**Comment by casperisfine** [comment]
> Do you want to wait for feedback from the other folks you tagged in the OP?

No don't worry, they're coworkers, I'll ping them more directly. And since I had to rebase again, I'd rather get this merged ASAP :p

**Comment by rafaelfranca** [comment]
Thanks for doing this :+1:

**Pull request merged**